docs/rustdoc/prehashing.md

```diff
@@ -0,0 +1,191 @@
+- If $n_p$ is too small, bins may not be sufficiently populated, reducing the chances of constructing valid tuples.
+
+### Example walkthrough
+- **Prover’s set** $S_p = \{A, B, C, D\}$
+- **Proof length** $u = 3$
+- **Search width** $d = 2$ (i.e., $t \in \{1, 2\}$)
+- **Hash functions**:
+  - $H_0(s)$: Prehash for binning elements of $S_p$.
+  - $H_1(v, t, s_1, \ldots, s_k)$: Determines the next bin to explore based on the sequence so far.
+  - $H_2(v, t, s_1, \ldots, s_u)$: Final validity check for the sequence.
+> ---
+> - The prover precomputes $H_0(s)$ for all elements in $S_p$, assigning each element to a bin:
+>   - $H_0(A) = 2$
+>   - $H_0(B) = 1$
+>   - $H_0(C) = 2$
+>   - $H_0(D) = 3$
+> - Bins are grouped as follows:
+>   - Bin 1: $\{B\}$
+>   - Bin 2: $\{A, C\}$
+>   - Bin 3: $\{D\}$
+> ---
+> - The prover starts with $t = 1$ and attempts to construct a valid sequence of length $u = 3$ using bounded DFS.
+> - Depth $k = 1$:
+>   - Compute $H_1(1)$ to determine the first bin to explore.
+>     - Assume $H_1(1) = 2$, so the prover looks in bin 2.
+>   - Select $s_1 = A$ from bin 2.
+>     - Compute $H_1(1, A)$ to check where to explore next.
+>     - $H_1(1, A)$ takes the current sequence $(1, A)$ and outputs the bin index for the next step.
+>     - Assume $H_1(1, A) = 3$, meaning the prover should explore bin 3.
+> - Depth $k = 2$:
+>   - Look in bin 3, which contains $\{D\}$.
+>   - Select $s_2 = D$.
+>   - Compute $H_1(1, A, D)$ to check where to explore next.
+>     - Assume $H_1(1, A, D) = 1$, meaning the prover should explore bin 1.
+> - Depth $k = 3$ (Final Element):
+>   - Look in bin 1, which contains $\{B\}$.
+>   - Select $s_3 = B$.
+>   - Compute $H_1(1, A, D, B)$ to check if this sequence can proceed further.
+>     - Assume $H_1(1, A, D, B) = 3$, which indicates a mismatch with the required condition. The sequence is invalid.
+> - Backtracking for $t = 1$:
+>   - After exhausting all possible sequences for $t = 1$, no valid proof is found.
+>   - The prover backtracks and tries other combinations of $s_1$, $s_2$, and $s_3$.
+> - Step 3: Retry with $t = 2$
+>   - The prover now starts the process for $t = 2$.
+> - Depth $k = 1$:
+>   - Compute $H_1(2)$ to determine the first bin.
+>     - Assume $H_1(2) = 1$, so the prover looks in bin 1.
+>   - Select $s_1 = B$.
+>   - Compute $H_1(2, B)$ to check where to explore next.
+>     - Assume $H_1(2, B) = 2$, meaning the prover should explore bin 2.
+> - Depth $k = 2$:
+>   - Look in bin 2, which contains $\{A, C\}$.
+>   - Select $s_2 = A$.
+>   - Compute $H_1(2, B, A)$ to check where to explore next.
+>     - Assume $H_1(2, B, A) = 3$, meaning the prover should explore bin 3.
+> - Depth $k = 3$ (Final Element):
+>   - Look in bin 3, which contains $\{D\}$.
+>   - Select $s_3 = D$.
+>   - Compute $H_1(2, B, A, D)$ to check if this sequence is valid so far.
```

Comment:
We shouldn't have this check (H1 is only used to extend), line 123 to 126 can be removed